packages/react-dom-bindings/src/server/ReactFizzConfigDOM.js

```diff
@@ -5711,5 +6034,79 @@ export function hoistResources(
+// with a postpone in the shell. In the context of a render we don't have a completed shell
+// if this is called before the shell finishes rendering which usually will happen anytime
+// anything suspends in the shell.
+export function emitEarlyPreloads(
+  renderState: RenderState,
+  resumableState: ResumableState,
+  shellComplete: boolean,
+): void {
+  const onHeaders = renderState.onHeaders;
+  if (onHeaders) {
+    const headers = renderState.headers;
+    if (headers) {
+      let linkHeader =
+        headers.preconnects + headers.fontPreloads + headers.highImagePreloads;
+
+      if (!shellComplete) {
+        // We use raw iterators because we want to be able to halt iteration
+        // We could refactor renderState to store these dually in arrays to
+        // make this more efficient at the cost of additional memory and
+        // write overhead. However this code only runs once per request so
+        // for now I consider this sufficient.
+        const queueIter = renderState.styles.values();
+        for (
+          let queueStep = queueIter.next();
+          headers.remainingCapacity > 0 && !queueStep.done;
+          queueStep = queueIter.next()
+        ) {
+          const sheets = queueStep.value.sheets;
+          const sheetIter = sheets.values();
+          for (
+            let sheetStep = sheetIter.next();
+            headers.remainingCapacity > 0 && !sheetStep.done;
+            sheetStep = sheetIter.next()
+          ) {
+            const sheet = sheetStep.value;
+            const props = sheet.props;
+            const key = getResourceKey(props.href);
+
+            renderState.resets.style[key] = PRELOAD_NO_CREDS;
+            const header = getStylesheetPreloadAsHeader(sheet);
+            linkHeader += header;
+            headers.remainingCapacity -= header.length;
+
+            // We already track that the resource exists in resumableState however
+            // if the resumableState resets because we postponed in the shell
+            // which is what is happening in this branch if we are prerendering
+            // then we will end up resetting the resumableState. When it resets we
+            // want to record the fact that this stylesheet was already preloaded
+            renderState.resets.style[key] =
+              typeof props.crossOrigin === 'string' ||
+              typeof props.integrity === 'string'
+                ? [props.crossOrigin, props.integrity]
+                : PRELOAD_NO_CREDS;
+          }
+        }
+      }
+      onHeaders({
+        // This is hacky but we always append headers into each queue assuming it is a subsequent
+        // header. This means that if the string has any length at all it will start with ", "
+        // so we truncate it here. This is ugly but the alternatives are to use arrays and join
+        // or to length check the queue each time we write to it. This is the most efficient option
```

Comment:
Well, this will always flatten the ropes and clone it I think so might not be so good. We should just check each time before we add. Since it's interned `if (linkHeader !== '') linkHeader += ', ';` might be fastest.

Reply:
ah that makes sense